presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.optimizations.PlanOptimizer;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class IterativeOptimizer
+        implements PlanOptimizer
+{
+    private final Set<Rule> rules;
+
+    public IterativeOptimizer(Set<Rule> rules)
+    {
+        this.rules = ImmutableSet.copyOf(rules);
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (!SystemSessionProperties.isNewOptimizerEnabled(session)) {
+            return plan;
+        }
+
+        Memo memo = new Memo(idAllocator, plan);
+
+        Lookup lookup = node -> {
+            if (node instanceof GroupReference) {
+                return memo.getNode(((GroupReference) node).getGroupId());
+            }
+
+            return node;
+        };
+
+        exploreGroup(memo, lookup, idAllocator, symbolAllocator, memo.getRootGroup());
+
+        return memo.extract();
+    }
+
+    private boolean exploreGroup(Memo memo, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, int group)
```

Comment:
Instead of passing session objects (memo, lookup, idAllocator, symbolAllocator) in all the methods I would extract separate class which could have these objects as fields